verify.js

```diff
@@ -86,92 +51,134 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
 
   var header = decodedToken.header;
 
-  if (!~options.algorithms.indexOf(header.alg)) {
-    return done(new JsonWebTokenError('invalid algorithm'));
+  if(header.alg ==='none' && decodedToken.signature) {
+    return done(new JsonWebTokenError('invalid token: unsigned but signature is present'));
   }
 
-  var valid;
+  var getSecret;
 
-  try {
-    valid = jws.verify(jwtString, header.alg, secretOrPublicKey);
-  } catch (e) {
-    return done(e);
+  if(typeof secretOrPublicKey === 'function') {
+    if(!callback) {
+      return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
```

Comment:
Not sure, maybe we shouldn't block this use case, the consumer should know about it (or say it in README).
I can imagine a system that loads the keys into an object on startup, even if their execution is sync they could take advantage of this function (they would have to call `callback(null, key)` but their execution is sync).
Thoughts?

Reply:
How should the caller know whether the callback is going to be async or not?

Reply:
It's their code, they should know it, having a callback it does not mean it is async, if you don't perform any async operation inside it would be just a normal function call, ie:
```
const keys = {
  kid1: 'a',
  kid2: 'b'
}

function getSecret(jwtHeaders, jwtClaims, callback) {
  callback(null, keys[jwtHeaders.kid])
}
```
That's a sync `getSecret` function. It's true that it could lead into issues if a future developer, seeing the signature with `callback`, makes it async (ie: calling an external resource). Not sure if we should block that use case for consumer safeness.

Reply:
That's not what I meant. It's kind of hard to explain. Let me give it another shot. From the perspective of the `decode` function, it is not known if `getSecret` is async or not. That is important to know, because the flow will be different in both cases.

Therefor, if you're using the `decode` with a `getSecret` callback, it makes logical sense that `decode` also needs a callback. In that way it will work in both cases.

It is true that it doesn't imply it is async. If one is sure `getSecret` isn't async, one could also do this:

```javascript
let payload;
decode(token, getSecret, (err, result) => {
  payload = result;
});
```

Not that I recommend it, but it is a possibility. I wouldn't do it that way, because if `getSecret` later becomes async it would fail. I hope I explained better this time.

Reply:
🤔  I think we are kind of in the same page, so you as a user would feel weird if you pass a: `getSecret(blah, callback)` to a `verify(token, getSecret)`. Is it? Probably you are right and we can add the constraint to only use the `getSecret` when the verify function is callback-ish. If we get people asking to open it for the non-callback-ish signature we can remove the error later, the other way around would be more difficult.
Fine, let's leave the error. 

By the way I've been calling it `getSecret` function, but you can call it `getKey`, `keying` or whatever you think it's more meaningful.